CMakeLists.txt

```diff
@@ -0,0 +1,127 @@
+cmake_minimum_required(VERSION 3.23)
+project(
+    cppfront
+    LANGUAGES CXX
+    VERSION 0.1.0
+)
+
+option(
+    CPPFRONT_INSTALL_RULES "Enable install rules for cppfront" "${PROJECT_IS_TOP_LEVEL}"
+)
+
+add_executable(cppfront source/cppfront.cpp)
+add_executable(cppfront::cppfront ALIAS cppfront)
+target_compile_features(cppfront PRIVATE cxx_std_20)
+target_sources(
+    cppfront
+    PRIVATE
+    FILE_SET HEADERS
+    BASE_DIRS source
+    FILES source/common.h source/lex.h source/load.h source/parse.h source/sema.h
+)
+
+add_library(cpp2util INTERFACE)
+add_library(cppfront::cpp2util ALIAS cpp2util)
+target_compile_features(cpp2util INTERFACE cxx_std_20)
+target_include_directories(
+    cpp2util INTERFACE "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
```

Comment:
Since you are going for making this project FC-able, you should make `SYSTEM` conditional, so warnings in cppfront's headers are ignored where supported.

```cmake
set(warning_guard "")
if(NOT PROJECT_IS_TOP_LEVEL)
  set(warning_guard SYSTEM)
endif()

target_include_directories(
    cpp2util ${warning_guard} INTERFACE "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
)
```

Reply:
I'm against using `SYSTEM` at all. It goes against "keeping us honest by compiling to C++". It also goes against the valuable trend of keeping/restoring user-facing warnings that occur within the library's implementation.

Reply:
`SYSTEM` would apply only when the project is pulled in via `add_subdirectory` (and by proxy FC). This is the default anyway when importing a package in CMake and you have to manually opt-out: https://cmake.org/cmake/help/latest/prop_tgt/NO_SYSTEM_FROM_IMPORTED.html  
This is just about having a consistent interface in all scenarios or at least approaching that as much as possible.

You can also introduce an option to control this behavior in case that is desired, so it can be turned off in all cases:

```cmake
set(warning_guard "")
if(NOT PROJECT_IS_TOP_LEVEL)
  option(
      CPPFRONT_INCLUDES_WITH_SYSTEM
      "Use SYSTEM modifier for cppfront's includes, disabling warnings"
      ON
  )
  mark_as_advanced(CPPFRONT_INCLUDES_WITH_SYSTEM)
  if(CPPFRONT_INCLUDES_WITH_SYSTEM)
    set(warning_guard SYSTEM)
  endif()
endif()
```

Reply:
> This is just about having a consistent interface in all scenarios or at least approaching that as much as possible.

So the scenario is `add_subdirectory` without a binary directory, such that cppfront's targets are not imported targets and so `SYSTEM` does not apply by default. Even FC specifies the binary directory. Is there really value in providing a consistent interface for such consumers?

Reply:
Binary directory has nothing to do with it. There is no difference between `add_subdirectory(src)` and `add_subdirectory(thirdy-party/foo)`, from CMake's perspective it's all _your_ code that is going to be built together.

`find_package` on the other hand conventionally executes scripts that create `IMPORTED` targets that reference already built artifacts. `IMPORTED` targets have a whole set of implications, such as their headers not belonging to you and warnings from those not being useful, since you can't do anything about them anyway.

Reply:
Ah, yes, you're right. I agree it's useful to have for consistency.

Reply:
> such as their headers not belonging to you and warnings from those not being useful, since you can't do anything about them anyway.

When that's really the case, the library should workaround that. For example, see https://github.com/mpusz/units/issues/203. That was inspired by some r/cpp posts about MS's STL restoring user-facing warnings. I'd have linked them then, but I lost their track.

Reply:
That's tangential to what I'm referring to. `SYSTEM` behavior prevents `std::duration<float>(1.0)` from warning on the conversion from `double` to `float`, if your implementation effectively does `float(1.0)`, as opposed to the mandated `static_cast<float>` which MS's STL moved away from in some places.

Reply:
To be clear, I want your second suggested solution which allows disabling `SYSTEM` behavior: https://github.com/hsutter/cppfront/pull/15#discussion_r981142370.

While I appreciate the consistency your suggestion brings, it should be better. The status-quo is for everyone to do it their own way, probably inconsistently, if they do it at all. I wish this simply were one of CMake's `<PROJECT-NAME>` [Variables that Change Behavior](https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html#id3). Isn't it in the purview of CMake to facilitate bridging the gap between `find_package` and `add_subdirectory`?

Until then, I'd appreciate having a (function in a) module which defines these for me.
```CMake
include(JEGPBridgeFindPackageAndAddSubdirectory)
# ...
if(JEGP_cppfront_INCLUDES_WITH_SYSTEM) # Probably unnecessary for a built-in.
# ...
```

Reply:
@friendlyanon @JohelEGP - I implemented this with an option called `CPPFRONT_NO_SYSTEM`.